### PR TITLE
Molecule generation model (GeoDiff)

### PR DESCRIPTION
**Pull request by natolambert**
Added a new model file to add functionality for [this paper](https://github.com/MinkaiXu/GeoDiff) that does molecule generation via diffusion.

Pretrained models available are for two tasks, `drugs` and `qm9`:
```python
model1 = DualEncoderEpsNetwork.from_pretrained("natolambert/geodiff-qm9-dualenc")
model1 = DualEncoderEpsNetwork.from_pretrained("natolambert/geodiff-drugs-dualenc")
```

Will work on additional examples for this model and update this PR.
Some todo items:
- [x] setup colab dependencies to run model: is [here](https://colab.research.google.com/drive/1pLYYWQhdLuv1q-JtEHGZybxp2RBF8gPs?usp=sharing)
- [x] add model tests
- [x] update [colab](https://colab.research.google.com/drive/1pLYYWQhdLuv1q-JtEHGZybxp2RBF8gPs?usp=sharing)
- [ ] add documentation for model
- [x] rename model to `MoleculeGNN`
- [x] add dependency checks for `torch_geometric<2`, `pytorch 1.8`, and `torch_scatter` (a recommended installation method is in the colab)

Some comments:
- the GNN implementations came from [ConfGF](https://github.com/DeepGraphLearning/ConfGF)
- an issue to open after release will be to figure out how to port the existing models to the new version of pytorch geometric. The author is not sure if one can copy the parameter dict or if re-training will be needed. This will ease some of the specific requirements for this model.

### CODE REVIEW

File: src/diffusers/models/dualencoder_gfn.py [hunk 1]
```diff
@@ -506,14 +503,28 @@ def sigmoid(x):
...
+    def __init__(
+        self,
+        hidden_dim,
+        num_convs,
+        num_convs_local,
+        cutoff,
+        mlp_act,
+        beta_schedule,
+        beta_start,
+        beta_end,
```
Comment (patrickvonplaten) [inline src/diffusers/models/dualencoder_gfn.py:515]: Those we should put in the schedulers IMO :-)

File: src/diffusers/models/dualencoder_gfn.py [hunk 2]
```diff
@@ -506,14 +503,28 @@ def sigmoid(x):
...
-    def __init__(self, hidden_dim, num_convs, num_convs_local, cutoff, mlp_act, beta_schedule, beta_start, beta_end, num_diffusion_timesteps, edge_order, edge_encoder, smooth_conv):
+    def __init__(
+        self,
+        hidden_dim,
+        num_convs,
+        num_convs_local,
+        cutoff,
+        mlp_act,
+        beta_schedule,
+        beta_start,
```
Comment (patrickvonplaten) [inline src/diffusers/models/dualencoder_gfn.py:514]: Same here, I think they should go into the schedulers :-)

**Comment by natolambert** [comment]
Doing more digging, the model architecture is based on [GIN](https://paperswithcode.com/method/gin) and [SchNet](https://paperswithcode.com/method/schnet) (common graph neural networks).

There are local and global parameters of the molecule, and different components use different GNNs.

File: tests/test_modeling_utils.py [hunk 3]
```diff
@@ -15,6 +15,7 @@
 
 import inspect
 import math
+import pdb
```
Comment (patrickvonplaten) [inline tests/test_modeling_utils.py:18]: we should be careful to always remove those before merging :-) Totally fine to keep them for testing though!

**Comment by patrickvonplaten** [comment]
Hey @natolambert,

This PR looks super cool. I think we can merge it quite quickly! Leaving some feedback directly in the code. Regarding the notebook, the API: In general, the notebook looks very nice IMO. If we have to install certain dependencies so be it! Regarding the model API I would suggest to change it a bit. E.g.:

```python
# generate geometry with model, then filter it
model_outputs = model.forward(batch, t)

# this model uses additional conditioning of the outputs depending on the current timestep
epsilon = model.get_residual(pos, sigmas[t], model_outputs)["sample"]
```

Do you think we can merge this into a single forward pass e.g. something like:

```python
epsilon = model(batch, t, sigma)["sample"]
```
?

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 4]
```diff
@@ -68,6 +68,13 @@ def __init__(
         elif beta_schedule == "squaredcos_cap_v2":
             # Glide cosine schedule
             self.betas = betas_for_alpha_bar(num_train_timesteps)
+        elif beta_schedule == "sigmoid":
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddpm.py:71]: Very cool! Nice to see that "vanilla" DDPM can be used that easily

File: src/diffusers/models/dualencoder_gfn.py [hunk 5]
```diff
@@ -0,0 +1,695 @@
...
+        return d_emb * edge_attr  # (num_edge, hidden)
+
+
+def assemble_atom_pair_feature(node_attr, edge_index, edge_attr):
+    h_row, h_col = node_attr[edge_index[0]], node_attr[edge_index[1]]
+    h_pair = torch.cat([h_row * h_col, edge_attr], dim=-1)  # (E, 2H)
+    return h_pair
+
+
+def generate_symmetric_edge_noise(num_nodes_per_graph, edge_index, edge2graph, device):
```
Comment (patrickvonplaten) [inline src/diffusers/models/dualencoder_gfn.py:288]: I this function used somewhere?

Comment (patrickvonplaten) [reply src/diffusers/models/dualencoder_gfn.py:288]: Also didn't see it being used in the google colab

File: src/diffusers/models/dualencoder_gfn.py [hunk 6]
```diff
@@ -0,0 +1,695 @@
...
+    norm = torch.norm(vec, dim=-1, p=2, keepdim=True)
+    denom = torch.where(norm > limit, limit / norm, torch.ones_like(norm))
+    return vec * denom
+
+
+def is_local_edge(edge_type):
+    return edge_type > 0
+
+
+def is_train_edge(edge_index, is_sidechain):
```
Comment (patrickvonplaten) [inline src/diffusers/models/dualencoder_gfn.py:690]: is this method being used?

File: src/diffusers/models/dualencoder_gfn.py [hunk 7]
```diff
@@ -0,0 +1,695 @@
...
+from torch_geometric.typing import Adj, OptPairTensor, OptTensor, Size
+from torch_geometric.utils import dense_to_sparse, to_dense_adj
+from torch_scatter import scatter_add, scatter_mean
+from torch_sparse import SparseTensor, coalesce
+
+from ..configuration_utils import ConfigMixin
+from ..modeling_utils import ModelMixin
+
+
+BOND_TYPES = {t: i for i, t in enumerate(BT.names.values())}
```
Comment (patrickvonplaten) [inline src/diffusers/models/dualencoder_gfn.py:21]: It seems like we only need the lenght of this dict - could we maybe instead hardcode this value here or have it in the model config instead of having to rely on a dependency for this? I'm not sure it's worth it to add an additional (slighly dangerous looking dependency) just to get the length of `BOND_TYPES` - what do you think?

File: src/diffusers/models/dualencoder_gfn.py [hunk 8]
```diff
@@ -0,0 +1,695 @@
...
+import torch
+import torch.nn.functional as F
+from torch import Tensor, nn
+from torch.nn import Embedding, Linear, Module, ModuleList, Sequential
+
+from rdkit.Chem.rdchem import BondType as BT
+from torch_geometric.nn import MessagePassing, radius, radius_graph
+from torch_geometric.typing import Adj, OptPairTensor, OptTensor, Size
+from torch_geometric.utils import dense_to_sparse, to_dense_adj
+from torch_scatter import scatter_add, scatter_mean
```
Comment (patrickvonplaten) [inline src/diffusers/models/dualencoder_gfn.py:14]: Is `scatter_mean` used?

File: src/diffusers/models/dualencoder_gfn.py [hunk 9]
```diff
@@ -0,0 +1,695 @@
...
+from typing import Callable, Dict, Union
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from torch import Tensor, nn
+from torch.nn import Embedding, Linear, Module, ModuleList, Sequential
+
+from rdkit.Chem.rdchem import BondType as BT
+from torch_geometric.nn import MessagePassing, radius, radius_graph
```
Comment (patrickvonplaten) [inline src/diffusers/models/dualencoder_gfn.py:11]: We need to make sure that this model is only imported if `torch_geometric` or `torch_scatter` is present. We should therefore adapt the same logic we did for the optional `transformers` import. 

Could you do the following:

1. Add a `is_torch_scatter_available` and a `is_torch_geometric_available` to https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/__init__.py (just like it has been done for other dependencies)
2. Import `DualEncoderEpsNetwork` only if both dependencies are available here: https://github.com/huggingface/diffusers/blob/89f2011cede1a899cc3f7e4b47ae1178d3a6e68c/src/diffusers/models/__init__.py#L21 (just like we do it for LDM in pipelines here: https://github.com/huggingface/diffusers/blob/89f2011cede1a899cc3f7e4b47ae1178d3a6e68c/src/diffusers/pipelines/__init__.py#L9
3. Ony import the model if the depedency is available in the main init as well: https://github.com/huggingface/diffusers/blob/89f2011cede1a899cc3f7e4b47ae1178d3a6e68c/src/diffusers/__init__.py#L26 - otherwise import the dummy class
4. Having written 3.) run `make fix-styles` to automatically create the dummy class

File: src/diffusers/models/dualencoder_gfn.py [hunk 10]
```diff
@@ -0,0 +1,695 @@
...
+    N = pos.size(0)
+    dd_dr = (1.0 / edge_length) * (pos[edge_index[0]] - pos[edge_index[1]])  # (E, 3)
+    score_pos = scatter_add(dd_dr * score_d, edge_index[0], dim=0, dim_size=N) + scatter_add(
+        -dd_dr * score_d, edge_index[1], dim=0, dim_size=N
+    )  # (N, 3)
+    return score_pos
+
+
+class DualEncoderEpsNetwork(ModelMixin, ConfigMixin):
+    def __init__(
```
Comment (patrickvonplaten) [inline src/diffusers/models/dualencoder_gfn.py:463]: Make sure to not forget the `@register_to_config` decorator here: https://github.com/huggingface/diffusers/blob/89f2011cede1a899cc3f7e4b47ae1178d3a6e68c/src/diffusers/models/unet_2d.py#L13

File: src/diffusers/models/dualencoder_gfn.py [hunk 11]
```diff
@@ -0,0 +1,695 @@
...
+def eq_transform(score_d, pos, edge_index, edge_length):
+    N = pos.size(0)
+    dd_dr = (1.0 / edge_length) * (pos[edge_index[0]] - pos[edge_index[1]])  # (E, 3)
+    score_pos = scatter_add(dd_dr * score_d, edge_index[0], dim=0, dim_size=N) + scatter_add(
+        -dd_dr * score_d, edge_index[1], dim=0, dim_size=N
+    )  # (N, 3)
+    return score_pos
+
+
+class DualEncoderEpsNetwork(ModelMixin, ConfigMixin):
```
Comment (patrickvonplaten) [inline src/diffusers/models/dualencoder_gfn.py:462]: Since this is the first graph network NN in this library, we should be extra careful with the naming. 
Is this a universally understandable name? Do you think other graph networks would also use this architecuter? Should we make the name more generic in this case? Can we link to a paper here that defined that model architecture?

**Comment by natolambert** [comment]
I'll work through your comments soon @patrickvonplaten.
CC the original author @MinkaiXu in case he has any time to look.

File: src/diffusers/schedulers/scheduling_ddpm.py (hunk 4 above)
Comment (natolambert) [reply src/diffusers/schedulers/scheduling_ddpm.py:71]: According the author the model performs better quantitatively with a different scheduler, but visually I noticed no difference.

File: src/diffusers/models/dualencoder_gfn.py (hunk 7 above)
Comment (natolambert) [reply src/diffusers/models/dualencoder_gfn.py]: Oh interesting. I think this is all only used in training or in a different mode than the colab I built, I can delete this and a lot of other code (according to my tests, I will verify in the colab later).

Comment (natolambert) [reply src/diffusers/models/dualencoder_gfn.py]: SO many deletions!

File: src/diffusers/models/dualencoder_gfn.py (hunk 11 above)
Comment (natolambert) [reply src/diffusers/models/dualencoder_gfn.py]: Link to original code is above now too, I followed up with those authors asking if their's was original. https://github.com/DeepGraphLearning/ConfGF

File: src/diffusers/models/dualencoder_gfn.py (hunk 9 above)
Comment (natolambert) [reply src/diffusers/models/dualencoder_gfn.py]: So the painpoint here is that is needs specific versions too, because this code was made before breaking changes in torch_geometric. Can I do similar functions for that?

File: src/diffusers/models/molecule_gnn.py [hunk 12]
```diff
@@ -0,0 +1,640 @@
...
+    def __init__(self):
+        super(ShiftedSoftplus, self).__init__()
+        self.shift = torch.log(torch.tensor(2.0)).item()
+
+    def forward(self, x):
+        return F.softplus(x) - self.shift
+
+
+class CFConv(MessagePassing):
+    def __init__(self, in_channels, out_channels, num_filters, nn, cutoff, smooth):
```
[the remark this answers is not in the document]
Comment (clefourrier) [reply src/diffusers/models/molecule_gnn.py:69]: Also, if you are redefining here CFConv from the SchNet paper, I think you should init the linear/shifted softplus/linear layers (what you put in nn) here instead of passing them as arguments, as it makes it hard to follow/find the logic of the paper in your code.

[the remark this answers is not in the document]
Comment (clefourrier) [reply src/diffusers/models/molecule_gnn.py:69]: (If you want to allow something more general, the default init for this layer should still be the original one from the paper)

File: src/diffusers/models/molecule_gnn.py [hunk 13]
```diff
@@ -0,0 +1,640 @@
...
+    def forward(self, x, edge_index, edge_length, edge_attr):
+        x = self.conv(x, edge_index, edge_length, edge_attr)
+        x = self.act(x)
+        x = self.lin(x)
+        return x
+
+
+class SchNetEncoder(Module):
+    def __init__(
+        self, hidden_channels=128, num_filters=128, num_interactions=6, edge_channels=100, cutoff=10.0, smooth=False
```
Comment (clefourrier) [inline src/diffusers/models/molecule_gnn.py:122]: Isn't paper default for `num_interactions` 3?

File: src/diffusers/models/molecule_gnn.py [hunk 14]
```diff
@@ -0,0 +1,640 @@
...
+        else:
+            h = z
+        for interaction in self.interactions:
+            h = h + interaction(h, edge_index, edge_length, edge_attr)
+
+        return h
+
+
+class GINEConv(MessagePassing):
+    def __init__(self, nn: Callable, eps: float = 0.0, train_eps: bool = False, activation="softplus", **kwargs):
```
Comment (clefourrier) [inline src/diffusers/models/molecule_gnn.py:151]: Since you are importing classes from torchgeometric, why are you redefining GINEConv instead of using the default version? (Could be worth some doc to explain what is different  - the activation function choice?)

File: src/diffusers/models/molecule_gnn.py [hunk 15]
```diff
@@ -0,0 +1,640 @@
...
+            hidden) edge_attr: edge feature tensor with shape (num_edge, hidden)
+        Output:
+            node_attr graph feature
+        """
+
+        node_attr = self.node_emb(z)  # (num_node, hidden)
+
+        hiddens = []
+        conv_input = node_attr  # (num_node, hidden)
+
```
Comment (clefourrier) [inline src/diffusers/models/molecule_gnn.py:236]: What is node_attr for?

File: src/diffusers/models/molecule_gnn.py [hunk 16]
```diff
@@ -0,0 +1,640 @@
...
+    return edge_index, edge_type
+
+
+def get_distance(pos, edge_index):
+    return (pos[edge_index[0]] - pos[edge_index[1]]).norm(dim=-1)
+
+
+def eq_transform(score_d, pos, edge_index, edge_length):
+    N = pos.size(0)
```
Comment (clefourrier) [inline src/diffusers/models/molecule_gnn.py:421]: Function name unclear

File: src/diffusers/models/molecule_gnn.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (natolambert) [reply src/diffusers/models/molecule_gnn.py:69]: Oh yeah I hadn't seen this `nn` usage until now, I really dislike that. It's very confusing (and here is clear how copy-pasted blocks of this code were).

I think this was the result of copying code from multiple files too.

[the remark this answers is not in the document]
Comment (natolambert) [reply src/diffusers/models/molecule_gnn.py:69]: RE the second point, I don't feel as strongly about it. It was taken directly from a few codebases back. In this case, the classes `CFConv` and `InteractionBlock` could effectively become one?

File: src/diffusers/models/molecule_gnn.py (hunk 13 above)
Comment (natolambert) [reply src/diffusers/models/molecule_gnn.py:122]: For the SchNet paper? No clue. This was copied from author.

File: src/diffusers/models/molecule_gnn.py (hunk 14 above)
Comment (natolambert) [reply src/diffusers/models/molecule_gnn.py:151]: Followed-up with author @MinkaiXu directly, and he got this from the implementation he built on. I can look a little more.

Comment (natolambert) [reply src/diffusers/models/molecule_gnn.py:151]: I looked at the [source](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/conv/gin_conv.html#GINEConv) and it's the same with the optional addition of an activation.

File: src/diffusers/models/molecule_gnn.py [hunk 17]
```diff
@@ -0,0 +1,640 @@
+# Model adapted from GeoDiff https://github.com/MinkaiXu/GeoDiff
```
[the remark this answers is not in the document]
Comment (natolambert) [reply src/diffusers/models/molecule_gnn.py:1]: @anton-l or @patil-suraj any comment on this?

File: src/diffusers/models/molecule_gnn.py (hunk 15 above)
Comment (natolambert) [reply src/diffusers/models/molecule_gnn.py:236]: Do you mean with respect to the model / application? Or just in this code? Both of them I have only intermediate understanding of.

Comment (clefourrier) [reply src/diffusers/models/molecule_gnn.py:245]: No, I just meant that you do 
```
node_attr = self.node_emb(z)
conv_attr = node_attr
```
why not directly
```
conv_input = self.node_emb(z)
```
?

File: src/diffusers/models/molecule_gnn.py (hunk 14 above)
Comment (clefourrier) [reply src/diffusers/models/molecule_gnn.py]: Worth adding it in the class doc IMO

File: src/diffusers/models/molecule_gnn.py (hunk 13 above)
Comment (clefourrier) [reply src/diffusers/models/molecule_gnn.py:125]: Ok, I guess we can close this comment then (it is 3 in the SchNet paper, I looked it up along your code yesterday, but it's not that important, and if the other model relies on 6 as default, might be easier to keep 6)

File: src/diffusers/models/molecule_gnn.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (clefourrier) [reply src/diffusers/models/molecule_gnn.py]: Well, if you open the SchNet paper, there is Figure 2 (I think) which describes the model arch. I think it could be interesting if you went through the figure along with your code: having CFConv and InteractionBlock separate is not a problem, but it would make CFConv easier to match with the paper it it defined the `nn` layers in directly. For further readers of this code, it will be clearer, IMO

**Comment by patrickvonplaten** [comment]
We should be able to add it now that things are calmer :partying_face:

**Comment by georgosgeorgos** [comment]
@natolambert @clefourrier @patrickvonplaten any plan to integrate this PR?

**Comment by MinkaiXu** [comment]
I took a quick look and everything is pretty good (Thank you all for your efforts!) 
Ping me if there is any part I need to take a deeper look before merging the PR :)

**Comment by natolambert** [comment]
@georgosgeorgos and @MinkaiXu, sorry for the delay on merging. We got a little distracted by Stable Diffusion. I'll plan on merging the updates on main to this + the notebook then we should be able to close the PR soon.

The colab should run in its current form! So if you look at that, happy to take comments!

**Comment by natolambert** [comment]
Fighting the stale bot! We haven't forgotten about this, and actually moved it up the priority list today. Soon!

**Comment by natolambert** [comment]
Alright, I kind of messed up that rebase so there are some repeated commits in the log. 
The changes are all here. This is much closer to release now.

Unfortunately, there are some breaking issues right now. The primary one is that `torch` 1.8 (which is required to use the version of `torch_geometric` this was built on, does not support `@torch.inference_mode`, so importing the `safety_checker.py` fails. 

This is something I don't really want to mess with. @patrickvonplaten or @MinkaiXu do you have any comments / suggestions to get by this (tagged for different reasons -- @patrickvonplaten  taught me the version checking on input and @MinkaiXu knows those versions of torch).

**Comment by MinkaiXu** [comment]
I'm not quite familiar with the whole process --- `safety_checker.py` is some mandatory step for `diffusers`? If so, is there a source code for this file?

**Comment by natolambert** [comment]
@MinkaiXu, @patrickvonplaten moved fast and removed it [here](https://github.com/huggingface/diffusers/commit/b35bac4d3b1af7e2389809f96e8ada11da6cc503). An interesting little difference I am not aware of :)

**Comment by natolambert** [comment]
@patrickvonplaten @anton-l @patil-suraj: so I realized the tests I implemented are not that useful because in order to test them you need:
- pytorch 1.8,
- torch_geometric 1.7.2

These are kind of a pain to install with `conda install -c rusty1s pytorch-geometric=1.7.2` after installing `pytorch==1.8` from source.

How should we think about integrating these tests?

**Comment by patil-suraj** [comment]
Why do we need to install pytorch from source ? Also is `torch_geometric` not available through `pip` ?

**Comment by natolambert** [comment]
@patil-suraj TLDR is re-implemented a research paper's code and it was made on a version before a lot of breaking changes.

I'm actually not sure it any version of `torch_geometric` is available on pip, I don't think so.

**Comment by patrickvonplaten** [comment]
Do we already have a working google colab for this model.
Also if a model is simply to difficult to implement, we might also only add the part that is easy to add to diffusers and for the rest just rely on an example

**Comment by natolambert** [comment]
Oh yeah I can just put it all in the colab and not port it into diffusers. May be easier.

**Comment by patrickvonplaten** [comment]
> 

sounds good!

**Comment by natolambert** [comment]
We closed this PR in favor of an colab-only solution unless someone has time to update the source model to the new versions of `torch_geometric` so it doesn't require very different dependencies!

**Comment by georgosgeorgos** [comment]
> Do we already have a working google colab for this model. Also if a model is simply to difficult to implement, we might also only add the part that is easy to add to diffusers and for the rest just rely on an example

@patrickvonplaten @natolambert where is the working colab with geodiff in the `diffusers` repo?

**Comment by patrickvonplaten** [comment]
Let's leave it open until we have a colab version. We can also make use of community PRs :-)

**Comment by natolambert** [comment]
Created a [PR](https://github.com/huggingface/notebooks/pull/246) in notebooks and the colab is [here](https://colab.research.google.com/drive/1pLYYWQhdLuv1q-JtEHGZybxp2RBF8gPs#scrollTo=mgQA_XN-XGY2) -- need to update the link once the PR in notebooks is merged.

**Comment by rish-16** [comment]
Hey! Thanks for adding GeoDiff into the pipeline :D 

I was thinking TorsionDiff ([https://arxiv.org/abs/2206.01729](https://arxiv.org/abs/2206.01729)) might be a better approach to molecule generation. It requires fewer diffusion steps and is much faster at conformational generation than GeoDiff!
